rgo/git.py

```diff
@@ -167,6 +167,54 @@ def describe(self, pkg=None):
         version = version.replace("-", "_")
         return version, release
 
+    def _check_output(self, cmd):
+        env = dict(os.environ)
+        env["LC_ALL"] = "C.UTF-8"
+        return subprocess.check_output(cmd, cwd=self.cwd, encoding="utf-8", env=env).strip().splitlines()
+
+    def get_describe_long(self, ref):
+        """
+        Run git describe --tags --long <ref>
+        Return the a tuple with the following data:
+          * desc: raw `git desribe --tags --long <ref>` output
+          * tag: last available tag in commits <= ref (parsed from desc)
+          * commits: number of commits between the tag and the ref (parsed from desc)
+          * hash: hash of the ref (parsed from desc)
+        """
+        cmd = ["git", "describe", "--tags", "--long", ref]
+        desc = self._check_output(cmd)[0].strip()
+
+        tag, commits, hash = desc.rsplit("-", 2)
+
+        # covert number of commits to integer
+        commits = int(commits)
+
+        # remove the 'g' prefix and keep only the hash
+        hash = hash[1:]
+
+        return tag, commits, hash
+
+    def get_rpm_changelog(self):
+        """
+        Generate RPM changelog records from git history between self.ref and the last tag.
+        If self.ref == last tag, generate a changelog between self.ref and the last but one tag.
+        """
+        # determine the latest tag name and number of commit between ref and the tag
+        ref_tag, ref_commits, _ = self.get_describe_long(self.ref)
+
+        if ref_commits == 0:
+            # ref is tagged, there are no commits going to the changelog
+            # generating an empty changelog is not cool, let's generate a changelog between 2 tags instead
+            tag_tag, _, _ = self.get_describe_long(ref_tag + "^")
```

Comment:
How about `from_tag` instead of `tag_tag`? :slightly_smiling_face:

Reply:
Simplified, using only `tag` now.